### PR TITLE
feat: implement popover

### PR DESCRIPTION
**Pull request by rabelloo**
## Purpose

Implement Popover, which will be used as a base component in both Select and Tooltip.

## Approach

Core Popover is just a placement container for arbitrary content.
It relies on a `position: relative` parent which contains its target and the Popover itself.

React Popover has the option to use Portals to avoid being hidden by parents' overflow.
If a Portal is used (by setting a `target` ref), it does not require the aforementioned parent anymore.

## Testing

Storybook.

## Risks

Screen aware repositioning coming in a later PR.

### CODE REVIEW

**Comment by rabelloo** [comment]
Depends on #869, will wait for it to get merged to `main` then change target.

**Comment by josokinas** [comment]
> Depends on #869, will wait for it to get merged to `main` then change target.

Feel free to merge that one in, already approved.

File: packages/core/src/components/popover/popover.scss
Comment (josokinas) [inline packages/core/src/components/popover/popover.scss:93]:
```suggestion
  transform: translate#{$axis}(-50%);
```

**Review by josokinas — APPROVED** [review]

File: packages/core/src/components/popover/popover.scss
Comment (rabelloo) [reply packages/core/src/components/popover/popover.scss:93]: So, that is Prettier doing its thing. Try applying that change locally then saving, you'll see 😅 

I suppose we could disable prettier in this line if you really hate how it looks.

Comment (josokinas) [reply packages/core/src/components/popover/popover.scss:93]: LOL, never mind... Maybe Prettier will sort itself at some point with some upgrade.

**Comment by rabelloo** [comment]
One other thing I forgot to mention, the `placement` is handled by a single class, e.g. `-top-center`.
It's theoretically two modifiers, one for `position` and another for `alignment`.
But they're treated as a single thing.

For `Button`, we do `.-action--primary` to combine two modifiers.

We could potentially split them, which would make the CSS a bit better and terser, e.g. `-top -center`.
It wouldn't be possible if we did `-top-left` as recommended in design, but it is with `start | end | center`.

Thoughts?

**Comment by josokinas** [comment]
> Thoughts?

I prefer as it is now.

Less freedom of combination - less chance of some edge case.

**Comment by rabelloo** [comment]
Cool. Should we do `.-top--center` as `Button` does?
I can't remember why we chose `--` there.

**Comment by josokinas** [comment]
> Cool. Should we do `.-top--center` as `Button` does?

OK, I see what you mean. Yes, I think better to follow the `Button` component.

**Comment by rabelloo** [comment]
If that's the case, should we:
1. Split them in the component API as well: `<Popover position="top" alignment="center" />`
2. Keep the API but replace the hyphen: `'top-center'.replace('-', '--')`
3. Make the component exactly the same as the class: `<Popover placement="top--center" />`

File: packages/react/src/components/popover/useOnClickOutside.spec.ts
Comment (markorodic) [inline packages/react/src/components/popover/useOnClickOutside.spec.ts:55]: The underscore convention denotes unused args?

File: packages/react/src/components/popover/useObserver.ts
Comment (markorodic) [inline packages/react/src/components/popover/useObserver.ts:23]: Not clear why we need a timeout here?

File: packages/react/src/components/popover/useObserver.ts
Comment (markorodic) [inline packages/react/src/components/popover/useObserver.ts:14]: Why are we using a ref here for the timeout rather than some state?

File: packages/react/src/components/popover/useObserver.ts
Comment (markorodic) [inline packages/react/src/components/popover/useObserver.ts:9]: Is this generic providing flexibility on the type element because all `HTMLElements` inherit from Element? This is a nice use of generics :)

File: packages/core/src/components/popover/popover.scss
Comment (markorodic) [inline packages/core/src/components/popover/popover.scss:84]: I like this solution in the end.

File: packages/core/src/components/popover/popover.scss
Comment (markorodic) [inline packages/core/src/components/popover/popover.scss:24]: Is this a common convention for CSS comments?

File: e2e/visual-regression/all.e2e.ts
Comment (markorodic) [inline e2e/visual-regression/all.e2e.ts:26]: Wondering whether this test could be more deterministic? Perhaps display a success state when the effects have finished running or the iFrame has loaded.

**Review by markorodic — APPROVED** [review]

File: packages/react/src/components/popover/useOnClickOutside.spec.ts
Comment (rabelloo) [reply packages/react/src/components/popover/useOnClickOutside.spec.ts]: Yup, avoids `eslint` complaining too. Any name starting with `_` it will ignore.

File: packages/react/src/components/popover/useObserver.ts
Comment (rabelloo) [reply packages/react/src/components/popover/useObserver.ts:23]: To debounce if set.
For example if `debounceTime === 300` and multiple events trigger in under 300ms, `onChange` will only be called once.
We ended up not using `debounceTime` yet but I suspect I'll need to, in order to fix a bug with the "auto-optimal-positioning" feature coming next.

File: packages/react/src/components/popover/useObserver.ts
Comment (rabelloo) [reply packages/react/src/components/popover/useObserver.ts:14]: If we used state, every time the `effect` was called we'd update it, triggering a re-run of the entire hook.
Furthermore, that state would need to be part of the `effect`'s dependencies, so we'd have to destroy the previous Observer and create a new one.

Using `ref` makes it so we prevent a lot of render cycles, while still having access to the latest timeout "id" which is all we want.

File: packages/react/src/components/popover/useObserver.ts
Comment (rabelloo) [reply packages/react/src/components/popover/useObserver.ts:9]: Correct. Observers require an `Element` only so we need not be more specific than that.

In this case the generic doesn't matter because we eat `T` but it would if the return type was T, for instance.
If we did not use the generic then, it'd squash the definition from the outside.

For example:
```ts
const buttonRef = useRef<HTMLButtonElement>(null);

const getCurrentWithGenerics = <T extends Element>(ref: RefObject<T>) =>
  ref.current;
const getCurrentWithPlainType = (ref: RefObject<Element>) => ref.current;

const noiceType = getCurrentWithGenerics(buttonRef);
// HTMLButtonElement | null

const sadType = getCurrentWithPlainType(buttonRef);
// Element | null
```

File: packages/core/src/components/popover/popover.scss
Comment (rabelloo) [reply packages/core/src/components/popover/popover.scss:84]: Thanks for helping me iterate on it, the Sass loop looked horrendous, as did the verbose one-by-one definitions 😂

File: packages/core/src/components/popover/popover.scss
Comment (rabelloo) [reply packages/core/src/components/popover/popover.scss:24]: Nope, that's me. We could wrap it in a mixin too I suppose, or use a different comment style.
Happy to hear suggestions if you have them.
I did not spend a lot of time coming up with this, I just wanted to convey where it begins and where it ends.

File: e2e/visual-regression/all.e2e.ts
Comment (rabelloo) [reply e2e/visual-regression/all.e2e.ts:26]: The reason we need this is for Popover's Observer to fire and reposition the anchor element.
I tried thinking of a deterministic way, but it'd be specific to the Popover story, and this spec goes over several other components too.
I couldn't query for an `.ods-popover-anchor` because is not every story that would have one.

Any ideas?

Comment (rabelloo) [reply e2e/visual-regression/all.e2e.ts]: Did some more testing, turns out the theme class is only applied to the iframe's body after one React cycle so I could remove the `cy.wait`.
If it becomes flaky later, we can think of an additional check to perform.

**Comment by josokinas** [comment]
> If that's the case, should we:
> 
> 1. Split them in the component API as well: `<Popover position="top" alignment="center" />`
> 2. Keep the API but replace the hyphen: `'top-center'.replace('-', '--')`
> 3. Make the component API exactly the same as the class: `<Popover placement="top--center" />`
> 4. Use another separation token in the component API:
>    `<Popover placement="top center" />` and `'top center'.replace(' ', '--')`

To best align with a Button component, I would prefer 1st approach.

Others, especially the 4th one, I do not like - e.g. would you expect to also be able to set "center top"? Which in that case would not work.

**Comment by rabelloo** [comment]
> would you expect to also be able to set "center top"?

No, they're typed with the format `${position}-${alignment}`.

The only disadvantage of option 1 is verbosity, but I suppose it makes it more explicit and clearer too, as well as easier to use half of the default, e.g.
`<Popover position="bottom" /> // alignment is default 'center' `

I'm not entirely convinced which option I like best, but option 1 aligns better with Button as you said, so I'm happy to give that a try for version 1.
Will refactor asap.

**Comment by rabelloo** [comment]
[Refactor done](https://github.com/onfido/castor/pull/878/commits/035a4a00f0e9017fa11aa0985882f45eb879e458)

API is now split `<Popover place="bottom" align="start" />`

Let me know if you like it, if so we can merge.

File: packages/react/src/components/popover/popover.tsx
Comment (josokinas) [inline packages/react/src/components/popover/popover.tsx:10]: Probably this PR should now include [`react-dom` as dependency](https://github.com/onfido/castor/pull/752/files#diff-1f344ac391eeecc21ec0f01fb07430a47f4b80d20485c125447d54c33c4bbfc4R33)?

File: packages/react/src/components/popover/popover.tsx
Comment (josokinas) [inline packages/react/src/components/popover/popover.tsx:21]: Curious, why did you choose verb-y `align` and `place` rather than noun-y `alignment` and `placement`?

Comment (rabelloo) [reply packages/react/src/components/popover/popover.tsx:21]: Mostly because it's terser, but also because they exist in CSS like `align-items` and `place-items`.
One could argue that `place` represents `justify + align` but we don't `justify` the popover.

Perhaps `position & align` works better, which is pretty much how Popover relates to CSS.
i.e. `position: top; align-items: start` (even though the implementation doesn't use Flex or Grid).

Happy to go with nouns too.
Naming is hard.

Comment (josokinas) [reply packages/react/src/components/popover/popover.tsx:21]: Your call on this 😏

File: packages/react/src/components/popover/popover.tsx
Comment (rabelloo) [reply packages/react/src/components/popover/popover.tsx:10]: Ah yes, thanks for the reminder.

File: packages/react/src/components/popover/popover.tsx
Comment (rabelloo) [reply packages/react/src/components/popover/popover.tsx:26]: Went with `position & align`.
There's still time to change before releasing.

File: packages/core/src/components/popover/popover.stories.ts
Comment (josokinas) [inline packages/core/src/components/popover/popover.stories.ts:17]: Because https://github.com/onfido/castor/pull/882 was merged, please could you switch here (and in other places) to [use `optionsToSummary`](https://github.com/onfido/castor/pull/882/files#diff-28c6eb9281509786c4eec4fd1c77134e899e182aa1abfcd90e8b4168d3fea16dR10).

Comment (rabelloo) [reply packages/core/src/components/popover/popover.stories.ts]: Done.

**Comment by josokinas** [comment]
Screenshot sizes are different, did you generate them locally via Docker?

I think try to re-generate, as something does not add up.

**Comment by rabelloo** [comment]
No, I downloaded from GitHub Actions, it's weird.

**Pull request merged**
